### PR TITLE
CUDA: add DGX Spark network expert/tensor parallelism

### PR DESCRIPTION
**Pull request by shankinson**
## Summary

- Add guarded two- and four-rank NCCL expert parallelism and tensor parallelism for supported DeepSeek V4 Flash 0731 and GLM 5.2 GGUF layouts on CUDA.
- Extend expert parallelism to the eight-rank DeepSeek V4 Pro 0813 IQ2/Q2 layout. Pro world size 8 is EP-only; TP8 is rejected with guidance to use EP8.
- Keep every rank in one mirrored session graph while partitioning routed experts and model-specific decode work. The default path preserves reference arithmetic; the aligned-artifact throughput path remains explicit and quality-tested.
- Wire scalar, decode-batch, and mixed prefill/decode lifecycles through `ds4`, `ds4-bench`, `ds4-server`, `ds4-agent`, `ds4-eval`, and the official-continuation scorer.
- Add protocol-v11 quant/layout preflight, guarded unified-memory admission, worker failure propagation, CUDA kernel/indexer fixes, expanded protocol and memory tests, documentation, and release QA gates.

## Related upstream issues

- #262: implements the requested two-node, 128 GiB-per-node CUDA tensor-parallel cluster path.
- #651: exposes network tensor-parallel roles through `ds4-server` and also wires the agent, eval, benchmark, and scorer frontends.
- #595: addresses CUDA cross-node TP for GLM 5.2; the issue's single-node CUDA `--ssd-streaming` prefill failure remains outside this PR.
- #469: provides the cross-DGX TP/EP mechanism discussed there; native continuous-batch throughput improvements remain follow-up work.
- #123: overlaps the networked four-Spark use case; the separate single-host multi-GPU path is unchanged.

## Supported paths and safety

- CUDA network transport is NCCL (`--transport auto` or `nccl`).
- DeepSeek V4 Flash 0731 supports EP2, TP2, EP4, and TP4 for the documented IQ2_XXS/Q2_K/Q8_0 tensor mix.
- DeepSeek V4 Pro 0813 supports EP8 for the validated IQ2/Q2 profile. Its 384 routed experts are split into eight contiguous 48-expert ranges. TP8 is intentionally unsupported.
- GLM 5.2 retains guarded EP2/TP2/EP4/TP4 support for its documented routed and dense tensor layouts.
- Every rank opens the complete GGUF. The file may be local to each host or exposed through one shared filesystem such as BeeGFS; only rank-local model residency and work are sharded.
- Protocol v11 exchanges the model family, quant profile, structural layout fingerprint, context, world size, and parallel mode before NCCL starts.
- The unified-memory admission plan accounts for rank-local GGUF spans, mandatory and optional aligned artifacts, shared workspace, ordered-reduction scratch, and per-session KV/private buffers. DGX Spark deployments retain 32 GiB of headroom by default.
- `DS4_CUDA_TP_FAST_ALIGNED_EXPERTS=1` remains an explicit, non-byte-identical throughput opt-in. Exact arithmetic is the default.
- External MTP/DSpark drafting, GLM MTP, directional steering, SSD streaming, unsupported routed layouts, pipeline layer roles, and unsupported topology/profile combinations are rejected before network execution.
- The network control protocol is intended for trusted cluster networks and is not authenticated or release-stable; every rank must run the same commit and model bytes.

## Validation commands

Core build and regression commands included:

```sh
make clean
make
make test
make cpu
make cuda-regression
make test-metal-session-batch
```

The repeated throughput sweep used the same model, prompt, context allocation, power setting, and background-load controls for each comparison:

```sh
./ds4-bench -m MODEL --prompt-file speed-bench/promessi_sposi.txt \
  --ctx-start 4096 --ctx-max 8192 --ctx-alloc 9000 \
  --step-incr 4096 --gen-tokens 128 --power 100 --csv RESULT.csv
```

Network runs added the matching coordinator/worker role, EP or TP mode, world size, NCCL transport, and fabric endpoint on every rank. BeeGFS runs used `DS4_CUDA_NO_DIRECT_IO=1` because O_DIRECT returns `Remote I/O` on this deployment.

### Builds and deterministic regressions

- Warning-free clean CPU build of all five frontends.
- Warning-free clean CUDA Spark builds on sparks5-8 (NVIDIA GB10 / `sm_121`). `make cuda-regression` passed.
- Full model-backed `make test` passed on spark5 with the Flash 0731 IQ2/Q2 GGUF, including long-context, tool-call, logprob vectors, local golden output, tensor equivalence, server, protocol, CLI, and sampling gates.
- Parser and deterministic suites passed: eval extractors, agent, server, layer packing (97/97), placement/memory accounting (101/101), GPU arguments, CLI/network validation (103/103), sampling, and the expanded protocol harness.
- The protocol harness covers 2/4/8 ranks, out-of-order connection, seven-worker fan-out, quant/layout/context/mode mismatch, failed preflight, failed ACK, batch and mixed-batch propagation, disconnect, and clean shutdown.
- Apple M3 Max (128 GiB, macOS 26.5.1): warning-free clean Metal build of all five frontends. Full `make test` passed against the Flash 0731 IQ2/Q2 GGUF; after the final compile-only warning guard, the clean build and Metal session oracle re-passed.
- Metal session coverage passed for 2/4/8/16 sessions plus native and serialized fallback paths with exact logits.
- Existing four-Spark GLM checks remain green: model-backed server, agent, eval, EP/TP trace parity, CUDA row-slice/kernel regressions, and the tracked 100-case continuation fixture.

### DeepSeek V4 Pro EP8 acceptance

- The 464,627,334,560-byte Pro 0813 model was pre-verified and opened concurrently from one shared BeeGFS path on all eight ranks. Every mount reported the same size and inode identity.
- Real EP8 preflight completed on all ranks and assigned eight gap-free 48-expert ranges. A 4K CLI generation completed and four repeated greedy requests were identical.
- CLI, server, agent, eval, benchmark, and official-score entry points were exercised.
- Model-backed server batches passed at widths 2, 4, 8, and 16, including concurrent, streaming, sampled, mixed-length, and mixed prefill/decode requests.
- Exact and fast-aligned modes each completed all 100 official cases (2,279 target tokens). Exact `avg_nll` was `3.550955083`; fast-aligned was `3.530710318` (-0.570%).
- Exact and fast-aligned scoring were admitted at 85.50 GiB required against an 89.69 GiB guarded budget. A real 32K-context, 16-session launch was rejected before artifact construction at 125.13 GiB required.
- Pro's 1,024-row indexer and wide IQ2 activation path are covered by CUDA tests for 1,024-of-2,048 selection, 1,024-wide indexed attention, grouped wide-index rows, and lookup-table initialization beyond the Flash-sized activation cache.

### Speed regression matrix

Network comparisons use original PR head `e3cd1c9` as the baseline because EP2/TP2/EP4/TP4 do not exist on `main`. All values below are staged percentage deltas at the 4K / 8K frontiers in default exact mode:

| Path | Prefill delta | Generation delta |
|---|---:|---:|
| EP2 | +7.1% / +0.5% | +0.6% / +0.7% |
| TP2 | +1.6% / +0.4% | +1.2% / +0.3% |
| EP4 | +18.8% / +0.1% | +1.0% / +0.6% |
| TP4 | -0.8% / -0.1% | -2.4% / -0.6% |

Additional inherited-backend comparisons:

- Single-node CUDA, `main` to staged: prefill +2.3% / +2.8%; generation -1.1% / -2.0%. This memory-safe comparison disabled routed IQ2/Q2 aligned artifacts and optional Q8 conversion caches identically on both revisions.
- Metal, `main` to staged, five-minute-cooldown A/B/A comparison: prefill -0.7% / -1.9%; generation +0.1% / +2.4%.
- EP8 has no pre-change implementation. Its absolute functional benchmark measured 20.99/20.46 prefill tokens/s and 2.97/2.47 generation tokens/s at 4K/8K; these are acceptance measurements, not follow-up performance claims.

### CODE REVIEW

**Comment by Dakine135** [comment]
I have 2x DGX Sparks I would love to get ds4 running on. I can test if it would help.

**Comment by shankinson** [comment]
> I have 2x DGX Sparks I would love to get ds4 running on. I can test if it would help.

The changes should work with that setup. I've got 4 DGX sparks connected via a 200gbps switch and it worked in both 2 node and 4 node trials. Let me know if you run into any issues with it.

**Comment by xxnuo** [comment]
What speed can DSv4 Flash achieve on 2x DGX?

**Comment by shankinson** [comment]
> What speed can DSv4 Flash achieve on 2x DGX?

| Configuration | 4K prefill | 4K generation | 8K prefill | 8K generation |
|---|---:|---:|---:|---:|
| Single Spark | 785.74 t/s | 14.30 t/s | 808.79 t/s | 14.04 t/s |
| TP2 exact | 201.78 t/s | 16.28 t/s | 485.42 t/s | 15.69 t/s |
| TP2 fast-aligned | 282.92 t/s | 16.12 t/s | 706.98 t/s | 15.82 t/s |
| TP4 exact | 382.96 t/s | 17.27 t/s | 576.78 t/s | 17.29 t/s |
| TP4 fast-aligned | 473.94 t/s | 18.25 t/s | 712.56 t/s | 17.97 t/s |

Test conditions: DeepSeek V4 Flash 0731 IQ2XXS/Q2_K GGUF, 128 generated tokens, `--power 100`, and an explicitly matched `--prefill-chunk 2048` on every participating rank.

Not as fast as running the model in a tuned vllm deployment.
